resources/aws/asg_stack.go

```diff
@@ -0,0 +1,206 @@
+	TemplateURL string
+	VPCID       string
+}
+
+// CreateOrFail creates the autoscaling group stack in Cloud Formation
+// or returns the error.
+func (s *ASGStack) CreateOrFail() error {
+	stackInput := &cloudformation.CreateStackInput{
+		StackName:        aws.String(s.Name),
+		TemplateURL:      aws.String(s.TemplateURL),
+		TimeoutInMinutes: aws.Int64(defaultTimeout),
+		Parameters: []*cloudformation.Parameter{
+			{
+				ParameterKey:   aws.String("ASGMaxSize"),
+				ParameterValue: aws.String(strconv.Itoa(s.ASGMaxSize)),
+			},
+			{
+				ParameterKey:   aws.String("ASGMinSize"),
+				ParameterValue: aws.String(strconv.Itoa(s.ASGMinSize)),
+			},
+			{
+				ParameterKey:   aws.String("AssociatePublicIPAddress"),
+				ParameterValue: aws.String(fmt.Sprintf("%t", s.AssociatePublicIPAddress)),
+			},
+			{
+				ParameterKey:   aws.String("AZ"),
+				ParameterValue: aws.String(s.AvailabilityZone),
+			},
+			{
+				ParameterKey:   aws.String("HealthCheckGracePeriod"),
+				ParameterValue: aws.String(strconv.Itoa(s.HealthCheckGracePeriod)),
+			},
+			{
+				ParameterKey:   aws.String("IAMInstanceProfileName"),
+				ParameterValue: aws.String(s.IAMInstanceProfileName),
+			},
+			{
+				ParameterKey:   aws.String("ImageID"),
+				ParameterValue: aws.String(s.ImageID),
+			},
+			{
+				ParameterKey:   aws.String("LoadBalancerName"),
+				ParameterValue: aws.String(s.LoadBalancerName),
+			},
+			{
+				ParameterKey:   aws.String("InstanceType"),
+				ParameterValue: aws.String(s.InstanceType),
+			},
+			{
+				ParameterKey:   aws.String("KeyName"),
+				ParameterValue: aws.String(s.KeyName),
+			},
+			{
+				ParameterKey:   aws.String("SecurityGroupID"),
+				ParameterValue: aws.String(s.SecurityGroupID),
+			},
+			{
+				ParameterKey:   aws.String("SmallCloudConfig"),
+				ParameterValue: aws.String(s.SmallCloudConfig),
+			},
+			{
+				ParameterKey:   aws.String("SubnetID"),
+				ParameterValue: aws.String(s.SubnetID),
+			},
+			{
+				ParameterKey:   aws.String("VPCID"),
+				ParameterValue: aws.String(s.VPCID),
+			},
+		},
+		Tags: []*cloudformation.Tag{
+			{
+				Key:   aws.String(tagKeyName),
+				Value: aws.String(s.Name),
+			},
+			{
+				Key:   aws.String(tagKeyCluster),
+				Value: aws.String(s.ClusterID),
+			},
+		},
+	}
+
+	if _, err := s.Client.CreateStack(stackInput); err != nil {
+		return microerror.Mask(err)
+	}
+
+	if err := s.Client.WaitUntilStackCreateComplete(&cloudformation.DescribeStacksInput{
+		StackName: aws.String(s.Name),
+	}); err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+// Update updates the autoscaling group stack in Cloud Formation.
+func (s *ASGStack) Update() error {
+	currentStack, err := s.findExisting()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	updateableParams := map[string]string{
+		"ASGMaxSize": strconv.Itoa(s.ASGMaxSize),
+		"ASGMinSize": strconv.Itoa(s.ASGMinSize),
+		"ImageID":    s.ImageID,
```

Comment:
All the updatable params are now constants because they are used multiple times. The params that are only used when creating the stack are left as strings. As they just replicate the names in the ASGStack struct.

Passing the long list of params in is a weak point of the design. I'll see if I can find a way to improve it.